### PR TITLE
fix measure bottlenecks

### PR DESCRIPTION
**Pull request by RaphaelS1**
Depends on https://github.com/xoopR/distr6/pull/296

### CODE REVIEW

**Comment by bblodfon** [comment]
Above commits may seem strange but I had to fix these for the new RCLL tests

File: R/MeasureSurvRCLL.R [hunk 1]
```diff
@@ -73,17 +73,53 @@ MeasureSurvRCLL = R6::R6Class("MeasureSurvRCLL",
...
-      } else if (all(event)) { # all uncensored
-        # pdf at outcome time (survived *this* long)
-        out[event] = diag(as.matrix(distr[event]$pdf(event_times)))
-      } else { # mix
-        out[event] = diag(as.matrix(distr[event]$pdf(event_times)))
-        out[!event] = diag(as.matrix(distr[!event]$survival(cens_times)))
+      # Bypass distr6 construction if underlying distr represented by array
+      if (inherits(prediction$data$distr, "array")) {
+        surv = prediction$data$distr
+        if (length(dim(surv)) == 3) {
```
Comment (bblodfon) [inline R/MeasureSurvRCLL.R:80]: we need to do this as in graf score when we have 3d survival array

File: R/MeasureSurvRCLL.R [hunk 2]
```diff
@@ -73,17 +73,53 @@ MeasureSurvRCLL = R6::R6Class("MeasureSurvRCLL",
...
+          # survival 3d array, extract median
+          surv = .ext_surv_mat(arr = surv, which.curve = 0.5)
+        }
+        times = as.numeric(colnames(surv))
+
+        if (any(!event)) {
+          if (sum(!event) == 1) { # fix subsetting issue in case of 1 censored
```
Comment (bblodfon) [inline R/MeasureSurvRCLL.R:87]: the `surv[!event, ]` in the case of 1 censored creates a vector, which `t` tranposes the other way than what we want

Comment (bblodfon) [reply R/MeasureSurvRCLL.R:87]: and we need to subset here

File: R/MeasureSurvRCLL.R [hunk 3]
```diff
@@ -73,17 +73,53 @@ MeasureSurvRCLL = R6::R6Class("MeasureSurvRCLL",
...
+            cdf = t(1 - surv)
+          } else {
+            cdf = t(1 - surv[!event, ])
+          }
+
+          out[!event] = diag(
```
Comment (bblodfon) [inline R/MeasureSurvRCLL.R:93]: the `t` after the `diag` is not needed, as with the trick here we have rectangular matrices (beofre the subsetting above, we didn't)

File: R/PredictionDataSurv.R [hunk 4]
```diff
@@ -132,7 +132,13 @@ filter_prediction_data.PredictionDataSurv = function(pdata, row_ids, ...) {
...
 
     if (testDistribution(distr)) { # distribution
-      pdata$distr = distr[keep]
+      ok = inherits(distr, c("VectorDistribution", "Matdist", "Arrdist")) &&
+           length(keep) > 1 # edge case: Arrdist(1xYxZ) and keep = FALSE
+      if (ok) {
+        pdata$distr = distr[keep] # we can subset row/samples like this
+      } else {
+        pdata$distr = base::switch(keep, distr) # one distribution only
+      }
```
Comment (RaphaelS1) [inline R/PredictionDataSurv.R]: @bblodfon can you tell me what's happening here?

File: R/PredictionSurv.R [hunk 5]
```diff
@@ -171,10 +171,12 @@ PredictionSurv = R6Class("PredictionSurv",
       }
     },
     .distrify_survarray = function(x) {
-      if (inherits(x, "array")) { # can be matrix as well
+      if (inherits(x, "array") && nrow(x) > 0) { # can be matrix as well
         # create Matdist or Arrdist (default => median curve)
         distr6::as.Distribution(1 - x, fun = "cdf",
           decorators = c("CoreStatistics", "ExoticStatistics"))
+      } else {
+        NULL
```
Comment (RaphaelS1) [inline R/PredictionSurv.R:179]: @bblodfon  What's this case?

File: R/MeasureSurvRCLL.R [hunk 6]
```diff
@@ -77,15 +77,34 @@ MeasureSurvRCLL = R6::R6Class("MeasureSurvRCLL",
       # Bypass distr6 construction if underlying distr represented by array
       if (inherits(prediction$data$distr, "array")) {
         surv = prediction$data$distr
+        if (length(dim(surv)) == 3) {
+          # survival 3d array, extract median
+          surv = .ext_surv_mat(arr = surv, which.curve = 0.5)
+        }
         times = as.numeric(colnames(surv))
-        pdf = distr6:::cdfpdf(1 - surv)
```
Comment (RaphaelS1) [inline R/MeasureSurvRCLL.R]: @bblodfon  This function automatically detects if surv is an array or matrix

File: R/MeasureSurvRCLL.R (hunk 1 above)
Comment (RaphaelS1) [reply R/MeasureSurvRCLL.R:80]: @bblodfon see comment below, the new functiion I made `cdfpdf` checks if it's an array or not

File: R/MeasureSurvRCLL.R (hunk 2 above)
Comment (RaphaelS1) [reply R/MeasureSurvRCLL.R:87]: Really don't understand this sorry

File: R/MeasureSurvRCLL.R [hunk 7]
```diff
@@ -77,15 +77,34 @@ MeasureSurvRCLL = R6::R6Class("MeasureSurvRCLL",
...
+          if (sum(event) == 1) { # fix subsetting issue in case of 1 event
+            pdf = t(pdf)
+          } else {
+            pdf = t(pdf[event, ])
+          }
+
+          out[event] = diag(
+            distr6:::C_Vec_WeightedDiscretePdf(event_times, times, pdf = pdf)
+          )
         }
```
Comment (RaphaelS1) [inline R/MeasureSurvRCLL.R]: not sure I understand this

File: R/MeasureSurvRCLL.R (hunk 2 above)
Comment (bblodfon) [reply R/MeasureSurvRCLL.R:87]: image

Comment (bblodfon) [reply R/MeasureSurvRCLL.R:87]: and the last result is the `cdf` that is going in `C_Vec_WeightedDiscreteCdf` => which should be `times` as rows and patients as `columns` (1 in this case)

Comment (bblodfon) [reply R/MeasureSurvRCLL.R:87]: Thanks for pointing it out - actually the code/if-condition is not complete => in the case of 1 censored but some events: `event = c(F,T,T,T)` the code takes the whole `t(1-surv)` which is wrong and I hadn't tested for this

File: R/MeasureSurvRCLL.R (hunk 1 above)
Comment (bblodfon) [reply R/MeasureSurvRCLL.R:80]: I know, the problem is that following the code's logic, what comes most surely afterwards (at least 1 censored obs in the test set) is `cdf = t(1 - surv)` and `t` with a 3d `surv` array doesn't work (you need `aperm` and differentiate between matrix and 3d array then)

File: R/MeasureSurvRCLL.R (hunk 6 above)
Comment (bblodfon) [reply R/MeasureSurvRCLL.R]: Detects yes, but does it convert to 2d matrix that we need for `C_Vec_WeightedDiscretePdf`? I think not, right? => https://github.com/xoopR/distr6/blob/main/R/transformers.R#L30

File: R/PredictionDataSurv.R (hunk 4 above)
Comment (bblodfon) [reply R/PredictionDataSurv.R]: I was trying to filter `Distribution` objects (like the prediction object from `surv.parametric`) since I noticed we don't do any filtering on those and I needed on the tests to compare RCLL old and new code kind-of. The above is the result of testing some edge cases. 

In general you can subset with `distr[keep]` when you have a `Matdist`, `Arrdist` or an `VectorDistribution` if you have two or more observations. The problem starts when 1 observation is left. `distr[keep]` with a `VectorDistribution` leaves a `Distribution` (e.g. `WeibullAFT`-something in `surv.parametric`), a `WeightedDiscrete()` for `Matdist`, and an `Arrdist` of dim `1x...` (it doesn't degenerate to a `Matdist`). And that's all fine, we still use `pdata$distr = distr[keep]` to do this. But then if you redo some more filtering (on that same specific row index or another to go to 0 observations), `keep` will be either `TRUE` or `FALSE`, and most of the above are not subsettable or don't work at all:

Comment (bblodfon) [reply R/PredictionDataSurv.R]: image
arrdist_logical_subsetting

Comment (bblodfon) [reply R/PredictionDataSurv.R]: Thanks for asking this - based on the above, we might want to make an `Arrdist` to a `Matdist` when only one observation is left by design in `distr6`?

File: R/MeasureSurvRCLL.R (hunk 7 above)
Comment (bblodfon) [reply R/MeasureSurvRCLL.R]: Same issue with the `cdf` equivalent code, let's see it in that thread.

File: R/PredictionSurv.R (hunk 5 above)
Comment (bblodfon) [reply R/PredictionSurv.R:179]: This is the result of me harmonizing the output of `p$distr` when you filter too much :) See tests [here](https://github.com/mlr-org/mlr3proba/blob/fix_rcll_bottleneck/tests/testthat/test_PredictionSurv.R#L213). 

So if you do `p$filter(42)` and `42` is not a `row_id` you end up with 0 observations. But if `p$data$distr` is a matrix it actually leaves a 0-row matrix with the time points as columnames with the `drop = FALSE` (https://github.com/mlr-org/mlr3proba/blob/fix_rcll_bottleneck/R/PredictionDataSurv.R#L144). On the other hand, if `p$data$distr` is a distribution, it will leave `NULL` with the use of `switch`, which makes sense: https://github.com/mlr-org/mlr3proba/blob/fix_rcll_bottleneck/R/PredictionDataSurv.R#L140.

File: R/MeasureSurvRCLL.R (hunk 7 above)
Comment (bblodfon) [reply R/MeasureSurvRCLL.R]: No also changed to accommodate another edge case kind-of

File: R/PredictionDataSurv.R (hunk 4 above)
Comment (bblodfon) [reply R/PredictionDataSurv.R]: @RaphaelS1 Do you have an opinion on this?

File: R/MeasureSurvDCalibration.R [hunk 8]
```diff
@@ -62,18 +65,36 @@ MeasureSurvDCalibration = R6Class("MeasureSurvDCalibration",
...
+          cdf = t(1 - surv), FALSE, FALSE))
       } else {
-        si = diag(prediction$distr$survival(prediction$truth[, 1L]))
+        distr = prediction$distr
+        if (inherits(distr, c("Matdist", "Arrdist"))) {
+          si = diag(distr$survival(true_times))
+        } else { # VectorDistribution or single Distribution, e.g. WeightDisc()
+          si = as.numeric(distr$survival(data = matrix(true_times, nrow = 1L)))
+        }
       }
```
Comment (bblodfon) [inline R/MeasureSurvDCalibration.R:93]: This one is from using `lrn('surv.parametric')` and filtering to one observation, `p$distr` is a `WeibullAFT()` which is not a `VectorDistribution`, so `diag(distr$survival(true_times))` returned 0 and later `NaN` in the score
